internal/controller/ai_gateway_route.go

```diff
@@ -211,44 +211,98 @@ func (c *AIGatewayRouteController) syncAIGatewayRoute(ctx context.Context, aiGat
 		return fmt.Errorf("failed to reconcile extension policy: %w", err)
 	}
 
-	// Check if the HTTPRoute exists.
 	c.logger.Info("syncing AIGatewayRoute", "namespace", aiGatewayRoute.Namespace, "name", aiGatewayRoute.Name)
-	var httpRoute gwapiv1.HTTPRoute
-	err = c.client.Get(ctx, client.ObjectKey{Name: aiGatewayRoute.Name, Namespace: aiGatewayRoute.Namespace}, &httpRoute)
-	existingRoute := err == nil
-	if apierrors.IsNotFound(err) {
-		// This means that this AIGatewayRoute is a new one.
-		httpRoute = gwapiv1.HTTPRoute{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      aiGatewayRoute.Name,
-				Namespace: aiGatewayRoute.Namespace,
-			},
-			Spec: gwapiv1.HTTPRouteSpec{},
+	// List all HTTPRoutes in the same namespace as this aiGatewayRoute.
+	var existingHTTPRoutList gwapiv1.HTTPRouteList
+	err = c.client.List(ctx, &existingHTTPRoutList, client.InNamespace(aiGatewayRoute.Namespace))
+	if err != nil {
+		return fmt.Errorf("failed to list HTTPRoutes in %s namespace: %w", aiGatewayRoute.Namespace, err)
+	}
+	// Filter HTTPRoute owned by this aiGatewayRoute.
+	existingHTTPRoutes := make(map[string]*gwapiv1.HTTPRoute)
+	for _, httpRoute := range existingHTTPRoutList.Items {
+		for _, ownerRef := range httpRoute.OwnerReferences {
+			if ownerRef.UID == aiGatewayRoute.UID {
+				existingHTTPRoutes[httpRoute.Name] = &httpRoute
+			}
 		}
-		if err = ctrlutil.SetControllerReference(aiGatewayRoute, &httpRoute, c.client.Scheme()); err != nil {
-			panic(fmt.Errorf("BUG: failed to set controller reference for HTTPRoute: %w", err))
+	}
+	// Find all deduplicated AIServiceBackends of this aiGatewayRoute.
+	backends := make(map[string]*aigv1a1.AIServiceBackend)
+	var defaultBackend *aigv1a1.AIServiceBackend
+	for _, rule := range aiGatewayRoute.Spec.Rules {
+		for _, br := range rule.BackendRefs {
+			// key is to match HTTPRoute's name which is <backendName>-<aiGatewayRouteName>.
+			key := fmt.Sprintf("%s-%s", br.Name, aiGatewayRoute.Name)
+			if _, ok := backends[key]; ok {
+				continue
+			}
+			var backend *aigv1a1.AIServiceBackend
+			backend, err = c.backend(ctx, aiGatewayRoute.Namespace, br.Name)
+			if err != nil {
+				return fmt.Errorf("AIServiceBackend %s not found", key)
+			}
+			backends[key] = backend
+			// Choose default backend whose name is alphabetically smallest.
+			if defaultBackend == nil || defaultBackend.Name > backend.Name {
+				defaultBackend = backend
+			}
 		}
-	} else if err != nil {
-		return fmt.Errorf("failed to get HTTPRoute: %w", err)
 	}
 
-	// Update the HTTPRoute with the new AIGatewayRoute.
-	if err = c.newHTTPRoute(ctx, &httpRoute, aiGatewayRoute); err != nil {
-		return fmt.Errorf("failed to construct a new HTTPRoute: %w", err)
+	// The naming convention of HTTRoute is <backendName>-<aiGatewayRouteName>,
+	// It is impossible that a HTTRRoute belongs to two AIGatewayRoute.
```

Comment:
Can you have two AIGatewayRoute in the same namespace?

Reply:
Yes, it is possible that there are two AIGatewayRoute in the same namespace, and they have some same backends. In this implementation we create HTTPRoute per backend for each AIGatewayRoute.